temporal-sdk/src/main/java/io/temporal/worker/tuning/SlotSupplier.java

```diff
@@ -0,0 +1,89 @@
+ * at once.
+ *
+ * @param <SI> The type of information that will be used to reserve a slot. The three info types are
+ *     {@link WorkflowSlotInfo}, {@link ActivitySlotInfo}, and {@link LocalActivitySlotInfo}.
+ */
+@Experimental
+public interface SlotSupplier<SI> {
+  /**
+   * This function is called before polling for new tasks. Your implementation should block until a
+   * slot is available, then return a permit to use that slot.
+   *
+   * @param ctx The context for slot reservation.
+   * @return A permit to use the slot which may be populated with your own data.
+   * @throws InterruptedException The worker may choose to interrupt the thread in order to cancel
+   *     the reservation, or during shutdown. You may perform cleanup, and then should rethrow the
+   *     exception.
+   */
+  SlotPermit reserveSlot(SlotReservationContext<SI> ctx) throws InterruptedException;
+
+  /**
+   * This function is called when trying to reserve slots for "eager" workflow and activity tasks.
+   * Eager tasks are those which are returned as a result of completing a workflow task, rather than
+   * from polling. Your implementation must not block, and If a slot is available, return a permit
+   * to use that slot.
+   *
+   * @param ctx The context for slot reservation.
+   * @return Maybe a permit to use the slot which may be populated with your own data.
+   */
+  Optional<SlotPermit> tryReserveSlot(SlotReservationContext<SI> ctx);
+
+  /**
+   * This function is called once a slot is actually being used to process some task, which may be
+   * some time after the slot was reserved originally. For example, if there is no work for a
+   * worker, a number of slots equal to the number of active pollers may already be reserved, but
+   * none of them are being used yet. This call should be non-blocking.
+   *
+   * @param info The information about the task that will now be processed
+   * @param permit The permit that this implementation gave out when a slot was reserved.
+   */
+  void markSlotUsed(SI info, SlotPermit permit);
+
+  /**
+   * This function is called once a permit is no longer needed. This could be because the task has
+   * finished, whether successfully or not, or because the slot was no longer needed (ex: the number
+   * of active pollers decreased). This call should be non-blocking.
+   *
+   * @param reason The reason the slot is being released.
+   * @param permit The permit that was given out when the slot was reserved.
+   */
+  void releaseSlot(SlotReleaseReason reason, SlotPermit permit);
+
+  /**
+   * Because we currently use thread pools to execute tasks, there must be *some* defined
+   * upper-limit on the size of the thread pool for each kind of task, and your implementation must
+   * provide this number. You must not hand out more permits than this number.
+   *
+   * @return the maximum number of slots that can ever be in use at one type for this slot type.
+   */
+  int maximumSlots();
```

Comment:
Hm I wonder now if we really need this? It is totally valid to have an unbounded threadpool in Java by setting `maximumPoolSize` to `Integer.MAX_VALUE`.

https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ThreadPoolExecutor.html

Reply:
Yeah I'm just a little worried that could be a big footgun? They can always return a huge number here and accomplish the same thing if they want to?

Reply:
>Yeah I'm just a little worried that could be a big footgun?

What is the footgun? 

>They can always return a huge number here and accomplish the same thing if they want to?

True, but why isn't that just the default then?

Reply:
The footgun is possibly way over-subscribing the CPU in the case of activity executions, for example. You can easily end up running way too many concurrently. Buuuuut, I suppose that's definitely not happening if they use our fixed size supplier, since it'll do the right thing, and if they implement their own, well, then it should be fairly clear to them not to do that, or at least is their responsibility.

So, yeah, I can see allowing it to default to something big.

Reply:
>The footgun is possibly way over-subscribing the CPU in the case of activity executions, for example.

That is a problem of handing out to many slots. Basically I don't think there is any technical reason we need this in Java so if we add it here  we should add it to all SDKs.